src/main/scala/io/iohk/ethereum/network/p2p/messages/PV63.scala

```diff
@@ -25,37 +28,174 @@ object PV63 {
+              val v = rlpDecode[ByteString](value)
+              val child = if (v.nonEmpty && v.length < 32)
+                Right(MptValue(v))
+              else
+                Left(MptHash(v))
+
+              MptExtension(ByteString(decoded), child)
+          }
+        case _ =>
+          throw new RuntimeException("Cannot decode NodeData")
+      }
+    }
+  }
+
+  trait MptNode
+
   object NodeData {
     implicit val rlpEndDec = new RLPEncoder[NodeData] with RLPDecoder[NodeData] {
       override def encode(obj: NodeData): RLPEncodeable = {
         import obj._
-        RLPList(values: _*)
+
+        RLPList(values.map {
+          case Left(node) => RLPValue(rlpEncode[MptNode](node))
+          case Right(byteValue) => RLPValue(byteValue.toArray[Byte])
+        }: _*)
       }
 
       override def decode(rlp: RLPEncodeable): NodeData = rlp match {
-        case rlpList: RLPList => NodeData(rlpList.items)
+        case rlpList: RLPList =>
+          NodeData(rlpList.items.map { e =>
+            Try {
+              val v = rawDecode(e: Array[Byte])
+              Left(MptNode.rlpEndDec.decode(v))
+            }.getOrElse(Right(ByteString(e: Array[Byte])))
+          })
         case _ => throw new RuntimeException("Cannot decode NodeData")
       }
     }
 
     val code: Int = Message.SubProtocolOffset + 0x0e
   }
 
-  case class NodeData(values: Seq[RLPEncodeable]) extends Message {
+  case class NodeData(values: Seq[Either[MptNode, ByteString]]) extends Message {
     override def code: Int = NodeData.code
+
+    override def toString: String = {
+      val v = values.map(v => v.fold(_.toString, b => s"BytString(${Hex.toHexString(b.toArray[Byte])})"))
+
+      s"""NodeData{
+         |values: $v
+         |}
+       """.stripMargin
+    }
   }
 
+  case class MptBranch(children: Seq[Either[MptHash, MptValue]], terminator: ByteString) extends MptNode {
+    require(children.length == 16, "MptBranch childHashes length have to be 16")
+
+    override def toString: String = {
+      val childrenString = children.map { e =>
+        e.fold(a => s"Hash(${Hex.toHexString(a.hash.toArray[Byte])})", b => s"Value(${Hex.toHexString(b.value.toArray[Byte])})")
+      }.mkString("(", ",\n", ")")
+
+      s"""MptBranch{
+         |children: $childrenString
+         |terminator: ${Hex.toHexString(terminator.toArray[Byte])}
+         |}
+       """.stripMargin
+    }
+  }
+
+  case class MptExtension(keyNibbles: ByteString, child: Either[MptHash, MptValue]) extends MptNode {
+    override def toString: String = {
+      s"""MptExtension{
+         |key nibbles: $keyNibbles
+         |key nibbles length: ${keyNibbles.length}
+         |childHash: ${child.fold(a => s"Hash(${Hex.toHexString(a.hash.toArray[Byte])})", b => s"Value(${Hex.toHexString(b.value.toArray[Byte])})")}
+         |}
+       """.stripMargin
+    }
+  }
+
+  case class MptLeaf(keyNibbles: ByteString, value: Either[Account, ByteString]) extends MptNode {
```

Comment:
for example for hash `72e55bd454f4d2caf57508b08c832d06d598eac31ee1d53b823a674bc1863452` geth node returns returns `MptBranch` with empty hashes, so it is this case
It means that those branches do not exist